ui/hooks/useMultichainBalances.ts

```diff
@@ -1,151 +1,245 @@
+
+  const nonEvmTokensWithFiatBalances = useMemo(() => {
+    // build TokenWithFiat for each asset
+    return assetIds.map((caipAssetId) => {
+      const [caipChainId, address] = caipAssetId.split('/');
+      const [type] = address.split(':');
+      return {
+        chainId: caipChainId as `${string}:${string}`,
+        symbol: assetMetadataById[caipAssetId]?.symbol ?? '',
+        address: address as `${string}:${string}`,
+        string: balancesByAssetId[caipAssetId]?.amount ?? '0',
+        balance: balancesByAssetId[caipAssetId]?.amount ?? '0',
+        decimals: assetMetadataById[caipAssetId]?.units[0]?.decimals,
+        image: assetMetadataById[caipAssetId]?.iconUrl ?? '',
+        type: type === 'token' ? AssetType.token : AssetType.native,
+        tokenFiatAmount: new BigNumber(
+          balancesByAssetId[caipAssetId]?.amount ?? '1',
+        )
+          .times(
+            assetRates?.[caipAssetId]?.rate ??
+              nativeRates?.[
+                assetMetadataById[caipAssetId]?.units[0]?.symbol.toLowerCase()
+              ]?.conversionRate ??
+              '1',
+          )
+          .toNumber(),
+      };
+    });
+  }, [assetMetadataById, assetRates, assetIds, balancesByAssetId, nativeRates]);
+
+  return nonEvmTokensWithFiatBalances;
+};
+
+// This hook is used to get the balances of all tokens and native tokens across all chains
+// This also returns the total fiat balances by chainId/caipChainId
+export const useMultichainBalances = () => {
+  // EVM data
+  const evmBalancesWithFiatByChainId = useEvmAssetsWithBalances();
+  // Non-EVM data
+  const nonEvmBalancesWithFiatByChainId = useNonEvmAssetsWithBalances();
+
+  // return TokenWithFiat sorted by fiat balance amount
+  const assetsWithBalance = useMemo(() => {
+    return [
+      ...evmBalancesWithFiatByChainId,
+      ...nonEvmBalancesWithFiatByChainId,
+    ].sort((a, b) => (b.tokenFiatAmount ?? 0) - (a.tokenFiatAmount ?? 0));
+  }, [evmBalancesWithFiatByChainId, nonEvmBalancesWithFiatByChainId]);
+
+  // return total fiat balances by chainId/caipChainId
   const balanceByChainId = useMemo(() => {
-    return assetsWithBalance.reduce(
-      (acc: Record<`0x${string}`, number>, { chainId, tokenFiatAmount }) => {
-        if (!acc[chainId]) {
-          acc[chainId] = 0;
-        }
-        if (tokenFiatAmount) {
-          acc[chainId] += tokenFiatAmount;
-        }
-        return acc;
-      },
-      {},
-    );
-  }, [assetsWithBalance]);
+    return [
+      ...evmBalancesWithFiatByChainId,
+      ...nonEvmBalancesWithFiatByChainId,
+    ].reduce((acc: Record<Hex | CaipChainId, number>, tokenWithBalanceData) => {
+      if (!acc[tokenWithBalanceData.chainId]) {
+        acc[tokenWithBalanceData.chainId] = 0;
+      }
+      acc[tokenWithBalanceData.chainId] +=
+        tokenWithBalanceData.tokenFiatAmount ?? 0;
+      return acc;
+    }, {});
+  }, [evmBalancesWithFiatByChainId, nonEvmBalancesWithFiatByChainId]);
```

Comment:
I think that this hook is a good idea, but it may be better suited as a selector? I believe that we can make a deepEqual selector here. If it's a selector, I don't think the `useMemo` will work (outside of React's context). Are these memoizations necessary?